### ISSUE TITLE
Problem with resolution of Union

### ISSUE THREAD

**Issue by VinceDT** [microsoft/pyright#1071]
**Describe the bug**
Type checker shows problem, where there should be none (because Union is not resolved correctly?)

**To Reproduce**
See code below

**Expected behavior**
The type checker should not report an error here.

**Screenshots or Code**
```python
from typing import Union, List

class Foo:
    pass

class FooBar:
    foobar: Union[Foo, List[int]]

    def __init__(self) -> None:
        # self.foobar = [1, 2, 3]
        self.foobar = Foo()


def bar(m: Foo):
    pass

if __name__ == "__main__":
    f = FooBar()
    bar(f.foobar)
```
![image](https://user-images.githubusercontent.com/10597499/94710600-4f202480-0347-11eb-9fb1-ff295095fe6e.png)

**VS Code extension or command-line**
Usage via Pylance v2020.9.6, vscode Version: 1.49.2

Settings
"""
    "python.languageServer": "Pylance",
    "python.analysis.typeCheckingMode": "basic",
    "python.analysis.diagnosticMode": "workspace",
"""

Or what am I missing?

Thanks!

**Comment by erictraut** [comment]
This is the correct behavior. You have declared that the value of variable `FooBar.foobar` can be one of two possible types — either a `Foo` or a `List[int]`. Then you are trying to pass this value to a function `bar` that accepts only type `Foo`. The type checker is correctly reporting a type violation here because `f.foobar` might contain a `List[int]`, which is not handled as an argument to `bar`.

**Comment by erictraut** [comment]
The error will go away if you add the following check to your code:
```python
    if isinstance(f.foobar, Foo):
        bar(f.foobar)
```

That works because pyright is able to do something called "type narrowing" using the code flow graph. In this example, it narrows the type of f.foobar from `Union[Foo, List[int]]` to just `Foo` within that `if` statement.
